sdk/core/Azure.Core/src/Shared/DiagnosticScope.cs

```diff
@@ -370,9 +378,56 @@ static ActivityExtensions()
         private static Func<object, bool>? ActivitySourceHasListenersMethod;
         private static Func<string, string?, ICollection<KeyValuePair<string, object>>?, object?>? CreateActivityLinkMethod;
         private static Func<ICollection<KeyValuePair<string,object>>?>? CreateTagsCollectionMethod;
-
+        private static Func<Activity, string, object?>? GetCustomPropertyMethod;
+        private static Action<Activity, string, object>? SetCustomPropertyMethod;
+        private static string AzureSdkScopeLabel = "az.sdk.scope";
         private static readonly ParameterExpression ActivityParameter = Expression.Parameter(typeof(Activity));
 
+        public static bool IsNestedScopeEnabled(this Activity activity)
+        {
+            if (GetCustomPropertyMethod == null)
+            {
+                var method = typeof(Activity).GetMethod("GetCustomProperty");
```

Comment:
Why do we need to use reflection? Is this method not available on all target platforms? If not, then does this mean the feature only works on .NET 5+?

Reply:
We use reflection here a lot for the following reason:
- many APIs are available since DiagnosticSource package version 5.0.0 (or sometimes 6.0). CustomProperties are one of these relatively new APIs
- DiagnsoticSource is OOB nuget package - you can install 5.0.0 on .NET 4.6. So we can't use .NET version as a heuristic
- Consumers of our telemetry require DiagnsoticSource 5.0+
  - ApplicationInsights .NET SDK started depending on it in Nov 2020 and versions which don't depend on it are deprecated.
  - OpenTelemetry has similar timeline and there was never a GA verion of it which depended on older DiagnosticSoruce

So de-facto absolute majority of consumers has DiagnosticSource 5.0.0+, regardless of their .NET version.